healthchecks: init at 2.2.1

phaer opened the pull request:
###### Description of changes

This packages [healthchecks](https://github.com/healthchecks/healthchecks), the django app that powers https://healthchecks.io/ , as well as some missing dependencies.

This PR supersedes #156923, as it includes segno as well as sphinx-paramlinks

- packages build fine
-  basic functionality works fine
- There's a basic module & nixos test, all modeled after the existing netbox service

Most real-world setups will still need to use settings to configure their actual notification services like mail, pushover, slack or whatever. One can now just use the configuration keys from [the official documentation](https://healthchecks.io/docs/self_hosted_configuration/) to set the right environment variables.
 
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

phaer committed:
python3Packages.segno: init at 1.5.2

phaer committed:
python3Packages.cron-descriptor: init at 1.2.24

phaer committed:
python3Packages.cronsim: init at 2.1

phaer committed:
maintainers: add phaer

SuperSandro2000 commented:
> Most real-world setups will still need to use extraConfig to configure their actual notification services like mail, pushover, slack or whatever. As there are [quite a few options here](https://healthchecks.io/docs/self_hosted_configuration/) I personally don't think it makes sense add NixOS options for each and every one. What do you think?

That should be a generic settings style configuration option.

SuperSandro2000 commented on pkgs/development/python-modules/cronsim/default.nix:
```diff
@@ -0,0 +1,28 @@
+    sha256 = "sha256-nwlSAbD+y0l9jyVSVShzWeC7nC5RZRD/kAhCi3Nd9xY=";
+  };
+
+  doCheck = true;
```
```suggestion
```
that is the default. we don't need to repeat it.

SuperSandro2000 commented on pkgs/development/python-modules/cron-descriptor/default.nix:
```diff
@@ -0,0 +1,35 @@
+    description = "A Python library that converts cron expressions into human readable strings.";
+    homepage = "https://github.com/Salamek/cron-descriptor";
+    license = licenses.mit;
+    platforms = platforms.unix;
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/cron-descriptor/default.nix:
```diff
@@ -0,0 +1,35 @@
+  ];
+
+  meta = with lib; {
+    description = "A Python library that converts cron expressions into human readable strings.";
```
```suggestion
    description = "Library that converts cron expressions into human readable strings";
```

SuperSandro2000 commented on pkgs/development/python-modules/cron-descriptor/default.nix:
```diff
@@ -0,0 +1,35 @@
+  };
+
+  checkInputs = [
+    pep8
+    flake8
+    pep8-naming
```
We don't care at all about linting tests and should not execute them because they easily break eg on flake8 updates.

but I saw that there are tests we should execute explicitly https://github.com/Salamek/cron-descriptor/tree/master/tests

phaer replied:
It's implicitly running "python setup.py test" which runs the tests you've linked as well. If i naively remove the linting test inputs, setuptools tries to download them during build time and fails ofc. So the easy way is to just let them run. An alternative could be to try running the tests with pytest to use the existing disableTests facilities in that hook.  Do you think not running linting-tests is worth the effort?

SuperSandro2000 replied:
Yes, it is worth the effort to patch them out. Linters should not cause mass rebuilds and updating eg flake8 should not break healthchecks because they added some new rule. This makes also pythonPackages maintenance a lot easier and gives us way less surprises.

phaer replied:
The linters weren't even run, just installed to keep setup.py happy :) But I agree that it doesn't make sense to keep them in if they aren't really used. So patched out now.

SuperSandro2000 commented on pkgs/development/python-modules/cron-descriptor/default.nix:
```diff
@@ -0,0 +1,35 @@
+  pname = "cron_descriptor";
+  version = "1.2.24";
+
+  # Fetch source from GitHub, because pypi releases do not contain tests.
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+    };
+
+    users.users.healthchecks = {
+      home = "${cfg.dataDir}";
```
```suggestion
      home = "${cfg.dataDir}";
      homeMode = "700";
```

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+        wantedBy = [ "healthchecks.target" ];
+        after = [ "healthchecks.service" ];
+
+        environment = {
+          PYTHONPATH = pkg.pythonPath;
+        };
```
This shouldn't be required. Are we missing the wrapping for the binary maybe?

phaer replied:
You are right, it wasn't required for most of the services. But it is required for the healthchecks service as that one isn't simply run from the wrapper. We start gunicorn, which loads our WSGI service as a python module and therefore needs to find it in PYTHONPATH. using `pkg.pythonPath` for gunicorns `--pythonpath` flag should work in theory but does not in practice, presumably because the path gets to long. (systemd complained about a bad unit file)

SuperSandro2000 replied:
🤔 I think we could avoid this but it would require a patch and creating another python script. Since that is more maintenance work we can set the env for that one service.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+        preStart = ''
+          ${pkg}/bin/healthchecks collectstatic --no-input
+          ${pkg}/bin/healthchecks remove_stale_contenttypes --no-input
+          ${pkg}/bin/healthchecks compress
```
That should be done at compile time in the package to only run it once.
Doesn't this also write static files somewhere which would be lost?

phaer replied:
But its behavior depends on settings which could be modified via `extraConfig`, so i think packages build-time is the wrong place to run it?

It writes files to STATIC_ROOT which is set to /var/lib/healthchecks/static by default, so should be persistent on most systems.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+        User = "healthchecks";
+        Group = "healthchecks";
+        StateDirectory = "healthchecks";
+        StateDirectoryMode = "0750";
```
```suggestion
        StateDirectoryMode = "0700";
```

phaer replied:
Why is it problematic to allow group access if we have defined a dedicated group for that service?

SuperSandro2000 replied:
Yeah, you're right. In that case it is alright.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+        Group = "healthchecks";
+        StateDirectory = "healthchecks";
+        StateDirectoryMode = "0750";
+        Restart = "on-failure";
```
```suggestion
        Restart = "always";
```

SuperSandro2000 replied:
I looked that up quickly. The only difference is that if healthchecks exits with status code 0 it would not restart. I think we don't want that.

https://www.freedesktop.org/software/systemd/man/systemd.service.html

phaer replied:
Right, thanks for looking it up!

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+    systemd.targets.healthchecks = {
+      description = "Target for all Healthchecks services";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target"];
```
```suggestion
      after = [ "network.target" "network-online.target" ];
```

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+      type = lib.types.bool;
+      default = false;
+      description = ''
+        A boolean that controls whether site visitors can create new accounts. Set it to False if you are setting up a private Healthchecks instance, but it needs to be publicly accessible (so, for example, your cloud services can send pings to it).
```
```suggestion
        A boolean that controls whether site visitors can create new accounts. Set it to False if you are setting up a private Healthchecks instance,
        but it needs to be publicly accessible (so, for example, your cloud services can send pings to it).
```

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+
+    listenAddress = mkOption {
+      type = types.str;
+      default = "[::1]";
```
Is this IPv4 and IPv6 at the same time?

phaer replied:
It's the loopback device, works with IPv4 as well.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+  '');
+in {
+  options.services.healthchecks = {
+    enable = mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = ''
+        Enable Healthchecks.
+      '';
+    };
```
This should use the enable option short form.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+      ln -s ${localSettingsFile} $out/opt/healthchecks/hc/local_settings.py
+    '';
+  }));
+  healthchecksManageScript = with pkgs; (writeScriptBin "healthchecks" ''
+    #!${stdenv.shell}
```
We can use writeShellScript to avoid adding the shebang ourselves.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+  healthchecksManageScript = with pkgs; (writeScriptBin "healthchecks" ''
+    #!${stdenv.shell}
+    export PYTHONPATH=${pkg.pythonPath}
+    sudo -u healthchecks ${pkg}/bin/healthchecks "$@"
```
```suggestion
    ${pkg}/bin/healthchecks "$@"
```
We should already be that user when executing

phaer replied:
In systemd services: yes, but it's also added to `environment.systemPackages` to run maintenance tasks manually and i think it's useful to do this with the right user by default, no?
(that's also copied from netbox, so we might want to adapt it there as well if so)

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+    '';
+  };
+  pkg = (pkgs.healthchecks.overrideAttrs (old: {
+    installPhase = old.installPhase + ''
+      ln -s ${localSettingsFile} $out/opt/healthchecks/hc/local_settings.py
+    '';
```
This would cause us to rebuild the package everytime. If we cannot give the config via an argument we should think about a wrapper.

phaer replied:
Hm.. this is how it's done in netbox atm, but that's a very valid concern. An alternative would be to patch settings.py to load local_settings from somewhere else, but for that we would need a path which is known at build time. Not sure what's the best way forward here, will investigate how other django apps in nixpkgs approach this problem.

phaer replied:
Another approach would be to create an extra directory in the nix store, containing both the default config and our changes, and put that in PYTHON_PATH before the default one, so effectively shadowing the one in the package. 
that would be similar to how nixos/modules/services/monitoring/graphite.nix does it. 

The thing is, we would need to patch BASE_DIR then in order to keep it working. Just patching settings.py to replace 
```
# Read additional configuration from hc/local_settings.py if it exists
if os.path.exists(os.path.join(BASE_DIR, "hc/local_settings.py")):
    from .local_settings import *
```
with a fixed but might be better, but again, we would need a pre-defined path for that. We could use something like /etc/healthchecks/settings.py for that. What do you think?

phaer replied:
@SuperSandro2000 I just implemented a solution using EnvironmentFile. The healthchecksManageScript wrapper became a bit more hacky and could probably be improved but it seems to work and configuration handling in general is much better than before IMO.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+      ALLOWED_HOSTS = ${builtins.toJSON cfg.allowedHosts}
+      # Default database engine is SQLite. So one can just check out code,
+      # install requirements.txt and do manage.py runserver and it works
+      DEFAULT_AUTO_FIELD = "django.db.models.AutoField"
+      DATABASES = {
+          "default": {
+          "ENGINE": "django.db.backends.sqlite3",
+          "NAME": "${cfg.dataDir}/healthchecks.sqlite",
+        }
+      }
```
Can I later overwrite this with postgres?

phaer replied:
Yes, anything in extraConfig will overwrite those default settings

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,216 @@
+          "NAME": "${cfg.dataDir}/healthchecks.sqlite",
+        }
+      }
+      with open("${cfg.secretKeyFile}", "r") as file:
+          SECRET_KEY = file.readline()
```
great, you can execute python in the settings 😐

phaer replied:
Yes, that's a standard django "feature". The line in question was copied from netbox https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/netbox.nix#L33

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -9982,6 +9982,19 @@
+      longkeyid = "rsa4096/0x00BCF29B32080700";
+      fingerprint = "5D69 CF04 B7BC 2BC1 A567  9267 00BC F29B 3208 0700";
+    }];
+  };
+
   philandstuff = {
```
```suggestion
  };
  philandstuff = {
```

SuperSandro2000 reviewed (commented):
We are using 2 spaces to indent nix files, this can easily be fixed with nixpkgs-fmt. 

Please add pythonImportsChecks to everything.

phaer commented on pkgs/servers/web-apps/healthchecks/default.nix:
```diff
@@ -0,0 +1,63 @@
+    ];
+
+    installPhase = ''
+      mkdir -p $out/opt/healthchecks
+      cp -r . $out/opt/healthchecks
```
I am open for other suggestions, but this is how I saw it done in other, similar, webapps in general?

SuperSandro2000 replied:
Things that are only presented to nginx sometimes also put files under plain $out for simplicity. We could use share but opt is fine I guess.

phaer commented on pkgs/servers/web-apps/healthchecks/default.nix:
```diff
@@ -0,0 +1,63 @@
+    meta = with lib; {
+      homepage = "https://github.com/healthchecks/healthchecks";
+      description = "A cron monitoring tool written in Python & Django ";
+      license = licenses.bsd3;
```
oh my, thanks for catching!

phaer commented:
> We are using 2 spaces to indent nix files, this can easily be fixed with nixpkgs-fmt.

Thanks, I adapted my files, but it seems like it's not yet consistent across nixpkgs? 

> Please add pythonImportsChecks to everything.

I think that's a default? `pythonImportsCheckPhase` seems to be run for all packages in my builds.

> That should be a generic settings style configuration option.

I think that would be easy with JSON, TOML, etc, but as the config is full python code in this case, I don't really think it's worth the effort to try writing that from nix? :)

SuperSandro2000 commented:
> Thanks, I adapted my files, but it seems like it's not yet consistent across nixpkgs?

In theory it should be. Scripts use 4 spaces, though.

> I think that's a default? pythonImportsCheckPhase seems to be run for all packages in my builds.

It is run but you need to supply it with the module name to import. We couldn't automate that since it often differs from pname.

SuperSandro2000 commented on pkgs/development/python-modules/sphinx-paramlinks/default.nix:
```diff
@@ -0,0 +1,31 @@
+  ];
+
+  # no tests
+  doCheck = false;
```
```suggestion
  doCheck = false;

  pythonImportsCheck = [ "sphinx_paramlinks" ];
```

SuperSandro2000 commented on pkgs/development/python-modules/segno/default.nix:
```diff
@@ -0,0 +1,41 @@
+    pytestCheckHook
+    pypng
+    pyzbar
+  ];
```
```suggestion
  ];

  pythonImportsCheck = [ "segno" ];
```

SuperSandro2000 commented on pkgs/development/python-modules/segno/default.nix:
```diff
@@ -0,0 +1,41 @@
+  };
+
+  propagatedBuildInputs = [
+    sphinx
+    sphinx-paramlinks
+    sphinx_rtd_theme
```
We only require those if we are building the docs. I am not sure if we do that but for that the should be in nativeBuildInputs.

phaer replied:
Ah good catch, yes! I've removed them - and sphix-paramlinks as it isn't used anymore.

SuperSandro2000 commented on pkgs/development/python-modules/cronsim/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+  checkInputs = [
+    pytestCheckHook
+  ];
```
```suggestion
  ];

  pythonImportsCheck = [ "cronsim" ];
```

SuperSandro2000 commented on pkgs/development/python-modules/cron-descriptor/default.nix:
```diff
@@ -0,0 +1,29 @@
+  preBuild = ''
+    sed -i "/'pep8\|flake8\|pep8-naming',/d" setup.py
+  '';
+
```
```suggestion

  checkPhase = ''
    ${python.interpreter} setup.py test
  '';

  pythonImportsCheck = [ "cron_descriptor" ];

```

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,184 @@
+      };
+
+    users.users.healthchecks = {
+      home = "${cfg.dataDir}";
+      homeMode = "700";
```
```suggestion
```
we don't need that with stateDirectory.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,184 @@
+      type = lib.types.bool;
+      default = false;
+      description = ''
+        A boolean that controls whether site visitors can create new accounts. Set it to False if you are setting up a private Healthchecks instance,
```
```suggestion
        A boolean that controls whether site visitors can create new accounts. Set it to false if you are setting up a private Healthchecks instance,
```

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,184 @@
+
+  healthchecksManageScript = with pkgs; (writeShellScriptBin "healthchecks" ''
+    sudo -u healthchecks sh <<EOF
+      export $(cat ${environmentFile} | xargs);
```
Not sure if we need quoting here.

phaer replied:
I think it's good enough for this use-case. Quoting values written to the environmentFile would be a good practice, but I am not sure if this simple service module is the right place for it. I've been surprised that there didn't seem to be a function to write an environmentFile with quoted values already in nixpkgs, but maybe I just missed it? :thinking:

phaer replied:
Asking in https://matrix.to/#/#dev:nixos.org resulted in a recommendation for https://skarnet.org/software/s6/s6-envdir.html . I could add that one if you'd like, but as said I personally think the current iteration works well enough.

SuperSandro2000 replied:
whatever you like

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,184 @@
+  healthchecksManageScript = with pkgs; (writeShellScriptBin "healthchecks" ''
+    sudo -u healthchecks sh <<EOF
+      export $(cat ${environmentFile} | xargs);
+      ${pkg}/bin/healthchecks $@
```
```suggestion
      ${pkg}/bin/healthchecks "$@"
```

phaer replied:
This wouldn't work as expected as multiple command line args to the script would be passed as a single string. e.g.

Current: 
```sh
$ healthchecks help check
usage: manage.py check [-h] [--tag TAGS] [--list-tags] [--deploy] [--fail-level {CRITICAL,ERROR,WARNING,INFO,DEBUG}] [--database DATABASES] [--version] [-v {0,1,2,3}] [--settings SETTINGS] [--pythonpath PYTHONPATH] [--traceback] [--no-color] [--force-color] [app_label ...]

Checks the entire Django project for potential problems.
[...]
```

With your change: 
```sh
$ healthchecks help check
Unknown command: 'help check'. Did you mean check?
Type 'manage.py help' for usage.
```

SuperSandro2000 replied:
Then they are quoting wrong beforehand already. I am suing that pattern all over the place in my dotfiles.

phaer replied:
Yes, the outer shell does or did. Just pushed a proposal which is even uglier, but allows us keeping the quotes. Not sure if that's an improved compared to the version without quotes, but it seems to work.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/915

SuperSandro2000 commented on pkgs/development/python-modules/cron-descriptor/default.nix:
```diff
@@ -0,0 +1,36 @@
+  };
+
+  # remove tests_require, as we don't do linting anyways
+  preBuild = ''
```
```suggestion
  postPatch = ''
```

SuperSandro2000 reviewed (commented):
Lets wait if someone responds to the discourse post otherwise we can merge this.

phaer committed:
healthchecks: init at 2.2.1

phaer commented:
> I just saw that we have 29 commits here. Can you please squash them to one per package, the maintainer addition and one for the service?

Yes, of course - wasn't sure it that would be squashed into one anyways.
But first got the order wrong, then saw your comment regarding preBuilt/postPatch ;)

healthchecks package uses the nixos test which uses the module, which uses the package. So should I split that in 3 commits, 2 commits, or leave it at one? With 2 commits the test declaration in the package would be broken, but that might not hurt as it's `passthru`?

SuperSandro2000 commented:
> > I just saw that we have 29 commits here. Can you please squash them to one per package, the maintainer addition and one for the service?
> 
> Yes, of course - wasn't sure it that would be squashed into one anyways. But first got the order wrong, then saw your comment regarding preBuilt/postPatch ;)
> 
> healthchecks package uses the nixos test which uses the module, which uses the package. So should I split that in 3 commits, 2 commits, or leave it at one? With 2 commits the test declaration in the package would be broken, but that might not hurt as it's `passthru`?

I didn't see that. 👀

aanderse commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,185 @@
+    ALLOWED_HOSTS = lib.concatStringsSep "," cfg.allowedHosts;
+    DEBUG = boolToPython cfg.debug;
+    REGISTRATION_OPEN = boolToPython cfg.registrationOpen;
+  } // cfg.settings;
```
This makes me think you want a set of [freeform](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules) module options instead.

phaer replied:
I did not know about them, thanks!

aanderse commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,185 @@
+in
+{
+  options.services.healthchecks = {
+    enable = mkEnableOption "healthchecks";
```
This would be a good spot to include the expectation that generally this software is to be run behind a reverse proxy. I didn't see any instructions in upstream documentation so presumably this is a trivial task and simply mentioning a reverse proxy is sufficient.

phaer replied:
I've added a short sentence, is that good?

aanderse replied:
LGTM :+1:

phaer commented:
@SuperSandro2000 Looks like you are using FIDO for authentication? Or do you just have it enabled on your hosts? Running `healthchecks-manage check` works in the test runner, as well as on one of my hosts.

I don't use FIDO yet and so I am unsure how to reproduce your issue locally. Does it work if you add `python3Packages.fido2` to `propagatedBuildInputs` of `healthchecks`? If so, i can test for the fido security option of course.

It's a much more open problem in general as healthchecks supports quite a few integrations which of course may introduce additional runtime-dependencies. Personally I believe supporting the core functionality out of the box and using overrides for integrations with runtime dependencies is at least a first good step. Or do we need options for all of them?

phaer commented:
Hi @SuperSandro2000, Hi @aanderse.

Thanks, that's all really great and helpful feed, I am learning a lot. I tried to address all of it, let me know if I am missing anything. Most notably the following to 3 things changed since your last reviews:

* I've added to dependencies, I had missed before, as they are not in requirements.txt but only in upstream Dockerfile
* I simplied the manage script wrapper. It's now the users obligation to call sudo, the wrapper exits with an error if not. This avoids my former subshell hackery and seems easier to maintain long-term.
* I went through the options and used a freeform submodule as suggested, and think it's a much cleaner implementation now.
  A small improvement could maybe be made, I there was a way to the naming scheme in settings from ENV_VAR to nixVariable, maybe with the options apply? :eyes: 

I've also added some more tests which should cover at least core functionality. Running the apps pre-populated repl via "healthchecks-manage shell" works now as well, It did not find a "python" executable in $PATH before.

I did have a test for user creation and was intending to write a test for logins and basic e2e tests for, well, healthcheck notifications, but CSRF tokens made it hard to do that with curl. And a full-blown selenium-based test could be a good project for another PR.  

Pending ofborg results, I think this could finally be mergable? What du you think?

(I've left the branch dirty for the moment, because I am not sure if you can easily review just the most recent changes if I squash and force-push. Will squash before eventual merge).

Thanks for your time and commitment, @SuperSandro2000 & @aanderse! :100: :tada:

phaer commented:
Ah, ofborg still fails. Going to check in the next few days, why exactly. Misinterpreted the error last time, looking at https://logs.nix.ci/?key=nixos/nixpkgs.175919&attempt_id=b70e7150-e264-45fa-bee5-0f132589b469 it seems like some device attestation related stuff or so breaks in the ci. The test passes locally for me :shrug:

SuperSandro2000 commented:
> Ah, ofborg still fails. Going to check in the next few days, why exactly. Misinterpreted the error last time, looking at [logs.nix.ci/?key=nixos/nixpkgs.175919&attempt_id=b70e7150-e264-45fa-bee5-0f132589b469](https://logs.nix.ci/?key=nixos/nixpkgs.175919&attempt_id=b70e7150-e264-45fa-bee5-0f132589b469) it seems like some device attestation related stuff or so breaks in the ci. The test passes locally for me shrug
> 
> EDIT: probably related to #176916 which ofborg already has, but don't locally.

Yeah, thats the log I posted earlier.

> (I've left the branch dirty for the moment, because I am not sure if you can easily review just the most recent changes if I squash and force-push. Will rebase on master & squash before eventual merge).

If you don't force push excessively GitHub can diff force pushes.

> And a full-blown selenium-based test could be a good project for another PR.

The tests should mostly prevent easy breakages while updating dependencies. In the end if no one is using the software it can break over time.

SuperSandro2000 commented on pkgs/servers/web-apps/healthchecks/default.nix:
```diff
@@ -36,6 +36,8 @@ py.pkgs.buildPythonApplication rec {
     segno
     cryptography
+    apprise
+    minio
```
We should probably sort that list alphabetically.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -19,10 +19,12 @@ let
+          echo "please run as user 'healtchecks'." >/dev/stderr
+          exit 1
+      fi
       export $(cat ${environmentFile} | xargs);
```
👍🏼

phaer commented:
As the fido package changed API and the new version now in nixpkgs master isn't supported by upstream yet, I think we've now took long enough for this PR that I would like to wait until upstream follow the update :smile:

SuperSandro2000 commented:
> As the fido package changed API and the new version now in nixpkgs master isn't supported by upstream yet, I think we've now took long enough for this PR that I would like to wait until upstream follow the update 😄

No, we don't have to. We have fancy package overwrites for things outside of pythonPackages https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#overriding-python-packages-overriding-python-packages

SuperSandro2000 commented on pkgs/servers/web-apps/healthchecks/default.nix:
```diff
@@ -0,0 +1,86 @@
+  py = python3.override {
+    packageOverrides = self: super: {
+      django = super.django_4;
+      fido2 = let
+        version = "0.9.3";
+      in super.fido2.overridePythonAttrs(old: {
+        inherit version;
+        src = super.fetchPypi {
+          pname = "fido2";
+          inherit version;
```
```suggestion
      fido2 = super.fido2.overridePythonAttrs(old: rec {
        version = "0.9.3";
        src = super.fetchPypi {
          pname = "fido2";
          inherit version;
```

phaer replied:
Ah, still not sure when rec is frowned upon and when not.

SuperSandro2000 replied:
Its fine to use rec here.

SuperSandro2000 commented on pkgs/servers/web-apps/healthchecks/default.nix:
```diff
@@ -0,0 +1,86 @@
+}:
+let
+  py = python3.override {
+    packageOverrides = self: super: {
```
```suggestion
    packageOverrides = final: prev: {
```
newer things use those better named variables.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,202 @@
+  cfg = config.services.healthchecks;
+  pkg = cfg.package;
+  boolToPython = b: if b then "True" else "False";
+  environment = cfg.settings // {
+    PYTHONPATH = pkg.pythonPath;
+    STATIC_ROOT = cfg.dataDir + "/static";
+    DB_NAME = "${cfg.dataDir}/healthchecks.sqlite";
+    ALLOWED_HOSTS = lib.concatStringsSep "," cfg.settings.ALLOWED_HOSTS;
+    DEBUG = boolToPython cfg.settings.DEBUG;
+    REGISTRATION_OPEN = boolToPython cfg.settings.REGISTRATION_OPEN;
+  };
```
```suggestion
  environment = {
    PYTHONPATH = pkg.pythonPath;
    STATIC_ROOT = cfg.dataDir + "/static";
    DB_NAME = "${cfg.dataDir}/healthchecks.sqlite";
    ALLOWED_HOSTS = lib.concatStringsSep "," cfg.settings.ALLOWED_HOSTS;
    DEBUG = boolToPython cfg.settings.DEBUG;
    REGISTRATION_OPEN = boolToPython cfg.settings.REGISTRATION_OPEN;
  } // cfg.settings;
```

I am not sure if we want to people to overwrite those settings or not.

aanderse replied:
If you set those down in `config` then they can be overwritten with `mkForce` which is my personal preference.

phaer replied:
I've applied the suggestion, cfg.settings overrides environment now.

The problem before was that i.e. settings.DEBUG (a boolean) would have overridden environment.DEBUG (a python string containing said boolean). 
I believe I solved that by using mkOptions `apply` for ALLOWED_HOSTS, DEBUG & REGISTRATION_OPEN.

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -0,0 +1,202 @@
+        };
+      };
+
+    users.users.healthchecks = {
+      isSystemUser = true;
+      group = "healthchecks";
+    };
+    users.groups.healthchecks = { };
+  };
```
This is currently not configurable but I am not sure if it must be.

aanderse replied:
Not required, though some users like this flexibility.

phaer replied:
I've made them configurable, hope that works?

aanderse commented:
I'll try to find a few minutes today to post exactly what I meant with my comments.

aanderse commented:
I took a look at your code and made a few commits so you could follow what I did and why: https://github.com/phaer/nixpkgs/compare/phaer-healthchecks...aanderse:nixos/healthchecks?expand=1

Please feel free to `cherry-pick` this directly into your branch and then `squash` - no need to attribute me on any of this. Note that I didn't try to build any of this... so I apologize in advance if anything doesn't work :sweat_smile:

phaer commented:
@aanderse Thanks for your work, really appreciated!

I didn't know about RFC052, going to drop the static UIDs. Will also pick your fixed user check in the manage script as well as the StateDirectory stuff.

Regarding configuration, I don't really see the benefit in this case, but I might be missing something? You can already override STATIC_ROOT or PYTHON_PATH by just setting settings.PYTHON_PATH to whatever, no mkForce needed.

phaer commented:
@aanderse @SuperSandro2000 So, I think that might be it, or are there any more issues which need to be handled right away?

From my perspective it seems good to merge. I believe the config is now flexible enough, but could of course be further improved should the need arise.

Of borg fails atm on aarch64-darwin as pyopenssl is marked broken, but passes on all other platforms.
